corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -635,6 +637,8 @@ def save_new_form(form):
                 attachment_writer.write()
                 for operation in operations:
                     operation.save()
+                for vault_item in vault_items:
+                    vault_item.save()
```

Comment:
since the vault items table is in the "default" django DB there is no transaction on that database here. Either the vault table should be sharded by the form ID or you'll need to add additional transaction handling for that DB. That would need to be added here as well as `FormProcessorSQL.save_processed_models`

Reply:
This should also delegate the saving to the vault api:
```
from corehq import vault

vault.save_values(vault_values)
```

Reply:
https://github.com/dimagi/commcare-hq/pull/28161/commits/32deb5f56bd68bcd9f432abaf0d0e4dcbbf75890